File: report/src/main/java/com/a/jgit/JGitTest.java
package com.a.jgit;

import com.a.jgit.diff.ClassesDiff;
import com.a.jgit.diff.classfiles.ClassMethodInfo;

import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.diff.DiffEntry;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.treewalk.AbstractTreeIterator;

import java.io.File;
import java.util.List;
import java.util.Set;

public class JGitTest {

    public static void main(String[] args)  throws Exception{
        String dir = "/Users/canglong/Documents/github_project/Android-Jacoco-Demo-builds";

        getDiffFileOfTwoBranch(dir, "build-2", "build-1");

        Set<ClassMethodInfo> methods = ClassesDiff.diffMethodsTwoBranch(dir, "build-2", "build-1");
        for (ClassMethodInfo method : methods) {
            System.out.println(method.toString());
        }

    }

   public static void getDiffFileOfTwoBranch(String gitDir , String newBranch,String  oldBranch ) throws Exception {
       Git git = Git.open(new File(gitDir));
       Repository repo = git.getRepository();
       AbstractTreeIterator oldTreeParser =ClassesDiff.prepareTreeParser(repo, "refs/heads/"+oldBranch);
       AbstractTreeIterator newTreeParser =ClassesDiff. prepareTreeParser(repo, "refs/heads/"+newBranch);

        // then the procelain diff-command returns a list of diff entries
        List<DiffEntry> diffs = git.diff().setOldTree(oldTreeParser).setNewTree(newTreeParser).call();
       for (DiffEntry diff : diffs) {
           System.out.println("diff:"+diff);
       }

    }
}

File: report/src/main/java/com/a/jgit/JGitTest.kt
package com.a.jgit

import com.a.jgit.diff.ClassesDiff
import com.a.jgit.diff.SourceCodeDiff
import org.eclipse.jgit.api.Git
import org.eclipse.jgit.lib.Ref
import org.eclipse.jgit.lib.Repository
import org.eclipse.jgit.revwalk.RevCommit
import org.eclipse.jgit.revwalk.RevTree
import org.eclipse.jgit.revwalk.RevWalk
import org.eclipse.jgit.treewalk.AbstractTreeIterator
import org.eclipse.jgit.treewalk.CanonicalTreeParser
import java.io.File
import java.io.IOException
import java.util.*

fun main(args: Array<String>) {
    val dir = "/Users/canglong/Documents/github_project/Android-Jacoco-Demo-builds"

    getDiffFileOfTwoBranch(dir, "build-2", "build-1")

    val classInfoList = SourceCodeDiff.diffMethodsTwoBranch(dir, "build-2", "build-1")
    println("-----------")
    classInfoList.forEach { it.methodInfos.forEach { println(it) } }
    println("-----------")
    val methods = ClassesDiff.diffMethodsTwoBranch(dir, "build-2", "build-1")
    methods.forEach { println(it) }

}

fun getDiffFileOfTwoBranch(gitDir: String, newBranch: String, oldBranch: String) {
    val git = Git.open(File(gitDir))
    val oldTreeParser = prepareTreeParser(git.repository, "refs/heads/${oldBranch}")
    val newTreeParser = prepareTreeParser(git.repository, "refs/heads/${newBranch}")

    // then the procelain diff-command returns a list of diff entries
    val diff = git.diff().setOldTree(oldTreeParser).setNewTree(newTreeParser).call()
    for (entry in diff) {
        println("Entry: $entry")
    }
}

@Throws(IOException::class)
private fun prepareTreeParser(repository: Repository, ref: String): AbstractTreeIterator? {
    // from the commit we can build the tree which allows us to construct the TreeParser
    val head: Ref = repository.exactRef(ref)
    RevWalk(repository).use { walk ->
        val commit: RevCommit = walk.parseCommit(head.getObjectId())
        val tree: RevTree = walk.parseTree(commit.tree.id)
        val treeParser = CanonicalTreeParser()
        repository.newObjectReader().use { reader -> treeParser.reset(reader, tree.id) }
        walk.dispose()
        return treeParser
    }
}